File: stacks/python/base/code/middleware.tpl.py
"""
File: middleware.tpl.py
Purpose: FastAPI middleware collection for auth, logging, rate limiting
Generated for: {{PROJECT_NAME}}
"""

import logging
import time
import uuid
from datetime import datetime
from typing import Callable, Optional

from fastapi import FastAPI, Request, Response, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp


# ============================================================================
# Request ID Middleware
# ============================================================================

class RequestIDMiddleware(BaseHTTPMiddleware):
    """Add unique request ID to each request."""
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        request_id = request.headers.get("X-Request-ID", str(uuid.uuid4()))
        
        # Store in request state
        request.state.request_id = request_id
        
        # Process request
        response = await call_next(request)
        
        # Add to response headers
        response.headers["X-Request-ID"] = request_id
        
        return response


# ============================================================================
# Logging Middleware
# ============================================================================

logger = logging.getLogger(__name__)


class LoggingMiddleware(BaseHTTPMiddleware):
    """Log all requests and responses."""
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        start_time = time.perf_counter()
        request_id = getattr(request.state, "request_id", "unknown")
        
        # Log request
        logger.info(
            f"Request started",
            extra={
                "request_id": request_id,
                "method": request.method,
                "path": request.url.path,
                "query": str(request.query_params),
                "client_ip": request.client.host if request.client else None,
                "user_agent": request.headers.get("user-agent"),
            }
        )
        
        # Process request
        response = await call_next(request)
        
        # Calculate duration
        duration = time.perf_counter() - start_time
        
        # Log response
        logger.info(
            f"Request completed",
            extra={
                "request_id": request_id,
                "method": request.method,
                "path": request.url.path,
                "status_code": response.status_code,
                "duration_ms": round(duration * 1000, 2),
            }
        )
        
        # Add timing header
        response.headers["X-Response-Time"] = f"{duration * 1000:.2f}ms"
        
        return response


# ============================================================================
# Error Handling Middleware
# ============================================================================

class ErrorHandlingMiddleware(BaseHTTPMiddleware):
    """Handle unhandled exceptions."""
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        try:
            return await call_next(request)
        except HTTPException:
            raise
        except Exception as e:
            request_id = getattr(request.state, "request_id", "unknown")
            
            logger.exception(
                f"Unhandled exception: {str(e)}",
                extra={"request_id": request_id}
            )
            
            # Return generic error response
            from fastapi.responses import JSONResponse
            return JSONResponse(
                status_code=500,
                content={
                    "success": False,
                    "errors": [
                        {
                            "code": "INTERNAL_ERROR",
                            "message": "An unexpected error occurred"
                        }
                    ],
                    "request_id": request_id,
                }
            )


# ============================================================================
# Rate Limiting Middleware
# ============================================================================

class RateLimitMiddleware(BaseHTTPMiddleware):
    """Rate limit requests by IP or user."""
    
    def __init__(
        self,
        app: ASGIApp,
        requests_per_minute: int = 60,
        redis_url: str = "redis://localhost:6379/0",
    ):
        super().__init__(app)
        self.requests_per_minute = requests_per_minute
        self.redis_url = redis_url
        self._redis = None
    
    async def get_redis(self):
        if self._redis is None:
            import redis.asyncio as redis
            self._redis = redis.from_url(self.redis_url)
        return self._redis
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        # Get client identifier
        client_id = self._get_client_id(request)
        
        # Check rate limit
        redis = await self.get_redis()
        key = f"ratelimit:{client_id}"
        
        current = await redis.get(key)
        if current is None:
            await redis.set(key, 1, ex=60)
            remaining = self.requests_per_minute - 1
        else:
            count = int(current)
            if count >= self.requests_per_minute:
                return self._rate_limit_response(request)
            await redis.incr(key)
            remaining = self.requests_per_minute - count - 1
        
        # Process request
        response = await call_next(request)
        
        # Add rate limit headers
        response.headers["X-RateLimit-Limit"] = str(self.requests_per_minute)
        response.headers["X-RateLimit-Remaining"] = str(max(0, remaining))
        
        return response
    
    def _get_client_id(self, request: Request) -> str:
        # Use user ID if authenticated, otherwise use IP
        user = getattr(request.state, "user", None)
        if user:
            return f"user:{user.id}"
        return f"ip:{request.client.host}"
    
    def _rate_limit_response(self, request: Request) -> Response:
        from fastapi.responses import JSONResponse
        return JSONResponse(
            status_code=429,
            content={
                "success": False,
                "errors": [
                    {
                        "code": "RATE_LIMITED",
                        "message": "Too many requests. Please try again later."
                    }
                ]
            },
            headers={
                "Retry-After": "60",
                "X-RateLimit-Limit": str(self.requests_per_minute),
                "X-RateLimit-Remaining": "0",
            }
        )


# ============================================================================
# Authentication Middleware
# ============================================================================

class AuthenticationMiddleware(BaseHTTPMiddleware):
    """Authenticate requests using JWT tokens."""
    
    PUBLIC_PATHS = {
        "/",
        "/health",
        "/health/db",
        "/health/redis",
        "/api/v1/auth/login",
        "/api/v1/auth/register",
        "/api/v1/auth/refresh",
        "/docs",
        "/redoc",
        "/openapi.json",
    }
    
    def __init__(self, app: ASGIApp, jwt_secret: str = "secret"):
        super().__init__(app)
        self.jwt_secret = jwt_secret
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        # Skip authentication for public paths
        if self._is_public_path(request.url.path):
            return await call_next(request)
        
        # Get token from header
        auth_header = request.headers.get("Authorization")
        if not auth_header or not auth_header.startswith("Bearer "):
            return self._unauthorized_response("Missing authentication token")
        
        token = auth_header[7:]
        
        try:
            # Validate token
            user = await self._validate_token(token)
            request.state.user = user
        except Exception as e:
            return self._unauthorized_response(str(e))
        
        return await call_next(request)
    
    def _is_public_path(self, path: str) -> bool:
        # Check exact match
        if path in self.PUBLIC_PATHS:
            return True
        # Check prefixes
        public_prefixes = ["/static/", "/api/v1/public/"]
        return any(path.startswith(p) for p in public_prefixes)
    
    async def _validate_token(self, token: str):
        """Validate JWT token and return user."""
        import jwt
        
        try:
            payload = jwt.decode(token, self.jwt_secret, algorithms=["HS256"])
            user_id = payload.get("sub")
            if not user_id:
                raise ValueError("Invalid token payload")
            
            # Fetch user from database
            # user = await get_user_by_id(user_id)
            # return user
            return {"id": user_id}
        except jwt.ExpiredSignatureError:
            raise ValueError("Token has expired")
        except jwt.InvalidTokenError:
            raise ValueError("Invalid token")
    
    def _unauthorized_response(self, message: str) -> Response:
        from fastapi.responses import JSONResponse
        return JSONResponse(
            status_code=401,
            content={
                "success": False,
                "errors": [
                    {"code": "UNAUTHORIZED", "message": message}
                ]
            }
        )


# ============================================================================
# Security Headers Middleware
# ============================================================================

class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Add security headers to responses."""
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        response = await call_next(request)
        
        # Add security headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"
        response.headers["Content-Security-Policy"] = "default-src 'self'"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        
        return response


# ============================================================================
# Middleware Configuration
# ============================================================================

def configure_middleware(app: FastAPI, settings: dict = None):
    """Configure all middleware for the application."""
    settings = settings or {}
    
    # CORS
    app.add_middleware(
        CORSMiddleware,
        allow_origins=settings.get("cors_origins", ["*"]),
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )
    
    # Security headers
    app.add_middleware(SecurityHeadersMiddleware)
    
    # Error handling
    app.add_middleware(ErrorHandlingMiddleware)
    
    # Logging
    app.add_middleware(LoggingMiddleware)
    
    # Request ID
    app.add_middleware(RequestIDMiddleware)
    
    # Rate limiting
    if settings.get("enable_rate_limiting", True):
        app.add_middleware(
            RateLimitMiddleware,
            requests_per_minute=settings.get("rate_limit", 60),
        )
    
    # Authentication (optional - can also use FastAPI dependencies)
    if settings.get("enable_auth_middleware", False):
        app.add_middleware(
            AuthenticationMiddleware,
            jwt_secret=settings.get("jwt_secret", "secret"),
        )
    
    return app

File: stacks/go/base/code/middleware.tpl.go
// File: middleware.tpl.go
// Purpose: HTTP middleware for common patterns
// Generated for: {{PROJECT_NAME}}

package middleware

import (
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/google/uuid"
)

// RequestID adds a unique request ID to context and response header
func RequestID(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestID := r.Header.Get("X-Request-ID")
		if requestID == "" {
			requestID = uuid.New().String()
		}
		w.Header().Set("X-Request-ID", requestID)
		next.ServeHTTP(w, r)
	})
}

// Logger logs request information with timing
func Logger(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		requestID := w.Header().Get("X-Request-ID")

		// Wrap response writer to capture status code
		wrapped := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}

		next.ServeHTTP(wrapped, r)

		duration := time.Since(start)
		log.Printf("[%s] %s %s %d %v",
			requestID, r.Method, r.URL.Path, wrapped.statusCode, duration)
	})
}

type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// RateLimiter provides simple in-memory rate limiting
type RateLimiter struct {
	mu       sync.RWMutex
	requests map[string][]time.Time
	limit    int
	window   time.Duration
}

func NewRateLimiter(limit int, window time.Duration) *RateLimiter {
	rl := &RateLimiter{
		requests: make(map[string][]time.Time),
		limit:    limit,
		window:   window,
	}

	// Cleanup goroutine
	go func() {
		ticker := time.NewTicker(window)
		defer ticker.Stop()
		for range ticker.C {
			rl.cleanup()
		}
	}()

	return rl
}

func (rl *RateLimiter) cleanup() {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	cutoff := time.Now().Add(-rl.window)
	for ip, timestamps := range rl.requests {
		var valid []time.Time
		for _, ts := range timestamps {
			if ts.After(cutoff) {
				valid = append(valid, ts)
			}
		}
		if len(valid) == 0 {
			delete(rl.requests, ip)
		} else {
			rl.requests[ip] = valid
		}
	}
}

func (rl *RateLimiter) Allow(ip string) bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	now := time.Now()
	cutoff := now.Add(-rl.window)

	// Filter old requests
	var valid []time.Time
	for _, ts := range rl.requests[ip] {
		if ts.After(cutoff) {
			valid = append(valid, ts)
		}
	}

	if len(valid) >= rl.limit {
		return false
	}

	rl.requests[ip] = append(valid, now)
	return true
}

func (rl *RateLimiter) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ip := r.RemoteAddr

		if !rl.Allow(ip) {
			http.Error(w, `{"error": "Rate limit exceeded"}`, http.StatusTooManyRequests)
			return
		}

		next.ServeHTTP(w, r)
	})
}

// Recover catches panics and returns 500 error
func Recover(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if err := recover(); err != nil {
				log.Printf("Panic: %v", err)
				http.Error(w, `{"error": "Internal server error"}`, http.StatusInternalServerError)
			}
		}()
		next.ServeHTTP(w, r)
	})
}

// SecurityHeaders adds security headers to responses
func SecurityHeaders(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("X-Content-Type-Options", "nosniff")
		w.Header().Set("X-Frame-Options", "DENY")
		w.Header().Set("X-XSS-Protection", "1; mode=block")
		w.Header().Set("Strict-Transport-Security", "max-age=31536000; includeSubDomains")
		next.ServeHTTP(w, r)
	})
}

// CORS middleware
func CORS(allowedOrigins []string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			origin := r.Header.Get("Origin")

			// Check if origin is allowed
			for _, allowed := range allowedOrigins {
				if allowed == "*" || allowed == origin {
					w.Header().Set("Access-Control-Allow-Origin", origin)
					w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
					w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")
					w.Header().Set("Access-Control-Max-Age", "86400")
					break
				}
			}

			if r.Method == "OPTIONS" {
				w.WriteHeader(http.StatusOK)
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

// Chain composes multiple middleware into a single handler
func Chain(middlewares ...func(http.Handler) http.Handler) func(http.Handler) http.Handler {
	return func(final http.Handler) http.Handler {
		for i := len(middlewares) - 1; i >= 0; i-- {
			final = middlewares[i](final)
		}
		return final
	}
}

// Usage:
// rateLimiter := NewRateLimiter(100, time.Minute)
// chain := Chain(
//     Recover,
//     RequestID,
//     Logger,
//     SecurityHeaders,
//     rateLimiter.Middleware,
// )
// http.Handle("/", chain(myHandler))
